gramps_webapi/auth/oidc.py

```diff
@@ -0,0 +1,360 @@
+        ROLE_MEMBER: current_app.config.get("OIDC_GROUP_MEMBER", ""),
+        ROLE_GUEST: current_app.config.get("OIDC_GROUP_GUEST", ""),
+    }
+
+    # Check if any role mapping is configured
+    has_role_mapping = any(group_name.strip() for group_name in role_mapping.values())
+    if not has_role_mapping:
+        logger.info("No OIDC role mapping configured (no GRAMPSWEB_OIDC_GROUP_* environment variables set). Preserving existing user roles.")
+        return None
+
+    # Extract user groups/roles from claims
+    user_groups = []
+
+    # Handle nested claims like 'realm_access.roles'
+    if '.' in role_claim:
+        claim_parts = role_claim.split('.')
+        claim_value = user_claims
+        for part in claim_parts:
+            claim_value = claim_value.get(part, {})
+        if isinstance(claim_value, list):
+            user_groups = claim_value
+    else:
+        # Handle direct claims like 'groups' or 'roles'
+        claim_value = user_claims.get(role_claim, [])
+        if isinstance(claim_value, list):
+            user_groups = claim_value
+        elif isinstance(claim_value, str):
+            user_groups = [claim_value]
+
+    # Fallback: if no groups found in claims, assign default guest role
+    if not user_groups:
+        logger.warning(f"No {role_claim} found in user claims. Available claims: {list(user_claims.keys())}. Assigning guest role.")
+        return ROLE_GUEST
+
+    highest_role = ROLE_GUEST
+
+    for role_level in sorted(role_mapping.keys(), reverse=True):
+        group_name = role_mapping[role_level]
+        if group_name and group_name in user_groups:
+            highest_role = role_level
+            break
+
+    logger.info(f"User {role_claim} {user_groups} mapped to role {highest_role}")
+    return highest_role
+
+
+def create_or_update_oidc_user(
+    userinfo: Dict,
+    tree: Optional[str] = None,
+    provider_id: str = "custom"
+) -> str:
+    """Create or update a user based on OIDC userinfo.
+
+    Args:
+        userinfo: User information from OIDC provider
+        tree: Tree identifier (optional)
+        provider_id: OIDC provider identifier
+
+    Returns the user GUID.
+    """
+    from flask import current_app
+
+    # Get provider-specific configuration
+    provider_config = get_provider_config(provider_id)
+    if not provider_config:
+        raise ValueError(f"Provider '{provider_id}' is not configured")
+
+    # Get username from provider-specific claim with fallback to 'sub'
+    username_claim = provider_config.get("username_claim", "preferred_username")
+    username = userinfo.get(username_claim) or userinfo.get("sub")
+
+    email = userinfo.get("email", "")
+    full_name = userinfo.get("name", "")
+
+    # For GitHub, get email from separate API call if not in userinfo
+    if provider_id == "github" and not email:
+        email = userinfo.get("email") or f"{userinfo.get('login', 'user')}@users.noreply.github.com"
```

Comment:
Will emails to @users.noreply.github.com be forwarded to the user?

Reply:
Good catch. No, they won't. Can only use this email for signing of commits on GitHub, my mistake. Removing this special case then.